scene/main/canvas_item.cpp

```diff
@@ -75,6 +103,34 @@ bool CanvasItem::is_visible_in_tree() const {
 	return true;
 }
 
+bool CanvasItem::is_visible_in_tree_with_cull_mask(int p_mask) const {
```

Comment:
My general feedback on this is that cull mask should not be verified here, you are going to get into a mess, plus you may have different views of the scene with different cull masks, which is really impossible to account for.

is_visible_in_tree() should only account for the visible property, not the cull masks.

Reply:
The idea with this part was to expose the necessary surface so that users could choose how they ask for visibility checks.

`is_visible_in_tree()` now automatically checks the viewport for what cull mask it is rendering.
`is_visible_in_tree_ignoring_cull_masks()` behaves as `is_visible_in_tree()` prior to the PR.
and `is_visible_in_tree_with_cull_mask(p_mask)` is exposed for clients in gdscript because as you say, it's impossible to account for how people will use it and what they expect.

I may have fudged the boundaries of what `is_visible_in_tree()` is for... I will admit. I didn't want to balloon the PR. Currently all of the UI/Control code expect visibility to coincide with rendering. It may be possible to sift through all of the UI code and replace each `is_visible_in_tree()` with a call that also checks the cull mask. Either that or controls on viewports that are culled won't behave correctly.

Reply:
I think its probably better to just add an `is_culled_by_viewport()` method and let the user use and combine as they wish.

Reply:
Ok, this is done. Originally I changed `is_visible_in_tree()` directly because it fixed nesting controls inside viewports and allowed hiding them as well as preventing them from being clicked and interacted with.

Now all controls and such that use `is_visible_in_tree()` need updated to `is_visible_to_viewport()` if that's what they really intended. The scope to fix this is too large for this PR, although the ground work is now in place to fix it.